packages/framework-provider-azure/src/library/events-store-adapter.ts

```diff
@@ -0,0 +1,140 @@
+import { BoosterConfig, EventEnvelope, NonPersistedEventEnvelope } from '@boostercloud/framework-types'
+import { CosmosClient, CreateOperationInput, JSONObject, OperationResponse, Response } from '@azure/cosmos'
+import { getLogger } from '@boostercloud/framework-common-helpers'
+import { partitionKeyForEvent } from './partition-keys'
+import { eventsStoreAttributes } from '../constants'
+
+type NonPersistedEventEnvelopePerPartitionKey = Record<string, Array<NonPersistedEventEnvelope>>
+
+const DEFAULT_CHUNK_SIZE = 100
+
+/**
+ * Limits: The Azure Cosmos DB request size limit constrains the size of the TransactionalBatch payload to not exceed 2 MB,
+ * and the maximum execution time is 5 seconds. There's a current limit of 100 operations per TransactionalBatch to ensure
+ * the performance is as expected and within SLAs.
+ *
+ * Errors: If there's a failure, the failed operation will have a status code of its corresponding error. All the other
+ * operations will have a 424 status code (failed dependency). The status code enables one to identify the cause of transaction failure.
+ *
+ * @param cosmosDb
+ * @param eventEnvelopes
+ * @param config
+ */
+export async function storeEvents(
+  cosmosDb: CosmosClient,
+  eventEnvelopes: Array<NonPersistedEventEnvelope>,
+  config: BoosterConfig
+): Promise<Array<EventEnvelope>> {
+  const logger = getLogger(config, 'store-events-adapter#storeEvents')
+  logger.debug('Storing EventEnvelopes with eventEnvelopes:', eventEnvelopes)
+  const envelopesWithCreatedAt: Array<EventEnvelope> = []
+
+  const eventsPerPartitionKey = eventEnvelopes.reduce(groupByPartitionKey, {})
+  for (const eventsPerPartitionKeyEntry of Object.entries(eventsPerPartitionKey)) {
+    const partitionKey = eventsPerPartitionKeyEntry[0]
+    const eventsInPartitionKey = eventsPerPartitionKeyEntry[1]
+    const chunksOfEventsInPartitionKey = chunkEvents(eventsInPartitionKey, DEFAULT_CHUNK_SIZE)
+    for (const eventListInChunk of chunksOfEventsInPartitionKey) {
+      const eventEnvelopesChunkWithCreatedAt = toEventEnvelopes(eventListInChunk)
+      envelopesWithCreatedAt.push(...eventEnvelopesChunkWithCreatedAt)
+      const inputOperations: Array<CreateOperationInput> = toInputOperations(eventEnvelopesChunkWithCreatedAt, config)
+      const batchResponse = await batchEvents(cosmosDb, config, inputOperations, partitionKey)
+
+      // Batch is transactional and will roll back all operations if one fails
+      if (batchResponse?.code !== 200) {
+        logger.error(batchResponse)
```

Comment:
Here it would be better if we prefix this log with something like `"An error ocurred storing a batch of events. Batch response: " and then the batch response.

Reply:
Expanded log message in commit 11bbb506.